Sources/NIOHTTPCompression/HTTPResponseDecompressor.swift

```diff
@@ -0,0 +1,197 @@
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOExtrasZlib
+import NIO
+import NIOHTTP1
+
+/// Specifies how to limit decompression inflation.
+public enum DecompressionLimit {
+    /// No limit will be set.
+    case none
+    /// Limit will be set on the request body size.
+    case size(Int)
+    /// Limit will be set on a ratio between compressed body size and decompressed result.
+    case ratio(Int)
+
+    func exceeded(compressed: Int, decompressed: Int) -> Bool {
+        switch self {
+        case .none:
+            return false
+        case .size(let allowed):
+            return compressed > allowed
+        case .ratio(let ratio):
+            return decompressed > compressed * ratio
+        }
+    }
+}
+
+public enum DecompressionError: Error {
+    case limit
+    case inflationError(Int32)
+    case initializationError(Int32)
+}
+
+public final class HTTPResponseDecompressor: ChannelDuplexHandler, RemovableChannelHandler {
+    public typealias InboundIn = HTTPClientResponsePart
+    public typealias InboundOut = HTTPClientResponsePart
+    public typealias OutboundIn = HTTPClientRequestPart
+    public typealias OutboundOut = HTTPClientRequestPart
+
+    private enum CompressionAlgorithm: String {
+        case gzip
+        case deflate
+    }
+
+    private enum State {
+        case empty
+        case compressed(CompressionAlgorithm, Int)
+    }
+
+    private let limit: DecompressionLimit
+    private var state = State.empty
+    private var stream = z_stream()
+    private var inflated = 0
+
+    public init(limit: DecompressionLimit) {
+        self.limit = limit
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let request = self.unwrapOutboundIn(data)
+        switch request {
+        case .head(var head):
+            if head.headers.contains(name: "Accept-Encoding") {
+                context.write(data, promise: promise)
+            } else {
+                head.headers.replaceOrAdd(name: "Accept-Encoding", value: "deflate, gzip")
+                context.write(self.wrapOutboundOut(.head(head)), promise: promise)
+            }
+        default:
+            context.write(data, promise: promise)
+        }
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            let algorithm: CompressionAlgorithm?
+            let contentType = head.headers[canonicalForm: "Content-Encoding"].first?.lowercased()
+            if contentType == "gzip" {
+                algorithm = .gzip
+            } else if contentType == "deflate" {
+                algorithm = .deflate
+            } else {
+                algorithm = nil
+            }
```

Comment:
maybe encapsulate into extension along the lines of `CompressionAlgorithm(header: head.headers[canonicalForm: "Content-Encoding"].first?.lowercased())`

Reply:
good idea, refactored, thanks!